pallets/passkey/src/lib.rs

```diff
@@ -124,56 +129,133 @@ pub mod module {
-		fn validate_signatures(call: &Call<T>) -> TransactionValidity {
-			match call {
-				Call::proxy { payload } => {
-					let signed_data = payload.passkey_public_key;
-					let signature = payload.passkey_call.account_ownership_proof.clone();
-					let signer = &payload.passkey_call.account_id;
-					match Self::check_account_signature(signer, &signed_data.into(), &signature) {
-						Ok(_) => Ok(ValidTransaction::default()),
-						Err(_e) => InvalidTransaction::BadSigner.into(),
-					}
-				},
-				_ => InvalidTransaction::Call.into(),
-			}
+#[derive(Encode, Decode, Clone, Eq, PartialEq, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+struct PasskeyNonce<T: Config>(pub T::Nonce);
+
+impl<T: Config> PasskeyNonce<T>
+where
+	<T as frame_system::Config>::RuntimeCall: Dispatchable<Info = DispatchInfo>,
+{
+	pub fn new(nonce: T::Nonce) -> Self {
+		Self(nonce)
+	}
+
+	pub fn validate(
+		&self,
+		who: &T::AccountId,
+		call: <T as frame_system::Config>::RuntimeCall,
+		info: &DispatchInfo,
+		len: usize,
+	) -> TransactionValidity {
+		let some = CheckNonce::<T>::from(self.0);
+		some.validate(who, &call, info, len)
+	}
+}
+
+impl<T: Config> Pallet<T>
+where
+	<T as frame_system::Config>::RuntimeCall: Dispatchable<Info = DispatchInfo>,
+{
+	fn filter_valid_calls(call: &Call<T>) -> Result<PasskeyPayload<T>, TransactionValidityError> {
+		match call {
+			Call::proxy { payload }
+				if T::PasskeyCallFilter::contains(&payload.clone().passkey_call.call) =>
+				return Ok(payload.clone()),
+			_ => return Err(InvalidTransaction::Call.into()),
 		}
+	}
+
+	fn validate_nonce(payload: &PasskeyPayload<T>) -> TransactionValidity {
+		let who = &payload.passkey_call.account_id.clone();
+		let nonce = payload.passkey_call.account_nonce;
 
-		/// Check the signature on passkey public key by the account id
-		/// Returns Ok(()) if the signature is valid
-		/// Returns Err(InvalidAccountSignature) if the signature is invalid
-		/// # Arguments
-		/// * `signer` - The account id of the signer
-		/// * `signed_data` - The signed data
-		/// * `signature` - The signature
-		/// # Return
-		/// * `Ok(())` if the signature is valid
-		/// * `Err(InvalidAccountSignature)` if the signature is invalid
-		fn check_account_signature(
-			signer: &T::AccountId,
-			signed_data: &Vec<u8>,
-			signature: &MultiSignature,
-		) -> DispatchResult {
-			let key: AccountId32 = T::ConvertIntoAccountId32::convert((*signer).clone());
-			let signed_payload: Vec<u8> = wrap_binary_data(signed_data.clone().into());
-
-			let verified = signature.verify(&signed_payload[..], &key);
-			if verified {
-				Ok(())
+		let some_call: &<T as Config>::RuntimeCall = &payload.passkey_call.call;
+		let info = &some_call.get_dispatch_info();
+
+		let passkey_nonce = PasskeyNonce::<T>::new(nonce);
+		passkey_nonce.validate(who, some_call.clone().into(), info, 0usize)
+	}
+
+	fn pre_dispatch_nonce(payload: &PasskeyPayload<T>) -> Result<(), TransactionValidityError> {
```

Comment:
I am considering whether we could also call pre_dispatch here. What are your thoughts on this?

Reply:
we should use the pre_dispatch from signed extension. that is one of the remaining things that I'm working on